libdnf/hy-query.c

```diff
@@ -1641,3 +1707,229 @@ hy_query_is_not_empty(HyQuery query)
+            }
+            s_available = pool_id2solvable(pool, id_available);
+            if ((s_installed->name == s_available->name) && (s_installed->evr == s_available->evr) && (s_installed->arch == s_available->arch)) {
+                matched = TRUE;
+                break;
+            }
+        }
+        if (!matched)
+            MAPSET(query->result, id_installed);
+    }
+    hy_query_free(query_installed);
+    hy_query_free(query_available);
+}
+
+void
+hy_filter_recent(HyQuery query, const long unsigned int recent_limit)
+{
+    Pool *pool = dnf_sack_get_pool(query->sack);
+    hy_query_apply(query);
+
+    for (Id id = 1; id < pool->nsolvables; ++id) {
+        if (!MAPTST(query->result, id)) {
+            continue;
+        }
+        DnfPackage *pkg = dnf_package_new(query->sack, id);
+        guint64 build_time = dnf_package_get_buildtime(pkg);
+        if (build_time <= recent_limit) {
+            MAPCLR(query->result, id);
+        }
+    }
+}
+
+static void
+add_duplicates_to_map(Pool *pool, Map *res, Queue samename, int start_block, int stop_block)
+{
+    Solvable *s_first, *s_second;
+    for (int pos = start_block; pos < stop_block; ++pos) {
+        Id id_first = samename.elements[pos];
+        s_first = pool->solvables + id_first;
+        for (int pos2 = pos + 1; pos2 < stop_block; ++pos2) {
+            Id id_second = samename.elements[pos2];
+            s_second = pool->solvables + id_second;
+            if ((s_first->evr == s_second->evr) && (s_first->arch != s_second->arch)) {
+                continue;
+            }
+            MAPSET(res, id_first);
+            MAPSET(res, id_second);
+        }
+    }
+}
+
+void
+hy_filter_duplicated(HyQuery query)
+{
+    Queue samename;
+    Pool *pool = dnf_sack_get_pool(query->sack);
+
+    hy_query_filter(query, HY_PKG_REPONAME, HY_EQ, HY_SYSTEM_REPO_NAME);
+    hy_query_apply(query);
+
+    hy_query_to_name_ordered_queue(query, &samename);
+
+    Solvable *considered, *highest = 0;
+    int start_block = -1;
+    int i;
+    MAPZERO(query->result);
+    for (i = 0; i < samename.count; ++i) {
+        Id p = samename.elements[i];
+        considered = pool->solvables + p;
+        if (!highest || highest->name != considered->name) {
+            /* start of a new block */
+            if (start_block == -1) {
+                highest = considered;
+                start_block = i;
+                continue;
+            }
+            if (start_block != i - 1) {
+                add_duplicates_to_map(pool, query->result, samename, start_block, i);
+            }
+            highest = considered;
+            start_block = i;
+            continue;
+        }
+    }
+    if ((start_block != -1) && (start_block < i)) {
+        add_duplicates_to_map(pool, query->result, samename, start_block, i);
+    }
+}
+
+int
+hy_filter_unneeded(HyQuery query, DnfSwdb *swdb, const int debug_solver)
+{
+    hy_query_apply(query);
+    GPtrArray *nevras = g_ptr_array_new();
```

Comment:
Tell to the array how to free the content
`GPtrArray *nevras = g_ptr_array_new_with_free_func(g_free);`
Or memory leak occurs on line 1898.